### ISSUE TITLE
Reading value of <Select>

### ISSUE THREAD

**Issue by Kinchkun** [kwebio/kweb-core#163]
Hi there,

I could not find way to observe the status from a `input(InputType.checkbox)`. Is there a way to bind a `KVar<Boolean>` to a checkbox status?

Thanks

**Comment by sanity** [comment]
Yes, an `InputElement` subclasses `ValueElement`, which has a `setValue(newValue: KVal<String>)` function, see [prelude.kt](https://github.com/kwebio/kweb-core/blob/master/src/main/kotlin/kweb/prelude.kt#L219).  You'll need to provide a `KVar<String>`, but that can be mapped from a `KVar<Boolean>`.

**Comment by sanity** [comment]
Closing for now, please reopen if this doesn't answer your question.

**Comment by shellderp** [comment]
Maybe I am missing something but setting `value` to true doesn't set `checked` to true, and I don't see any other way to access the `checked` attribute. The `checked()` method sets it but doesn't retrieve the value

**Comment by kernitus** [comment]
The issue here is that `value` doesn't reflect the status of the checkbox, but rather the presence of the `checked` attribute does. KWeb doesn't seem to have a way to get attributes easily. I have tried with executing some javascript code `withCallback`/`withResult`, but getting the element by id and then the attribute always returned null. For now I have worked around this by having a local variable, and forcefully setting the state of the checkbox everytime it is clicked, whilst also updating the variable.
```
                            var isChecked = false

                            checkbox.on.click {
                                isChecked = !isChecked
                                checkbox.checked(isChecked)
                            }
```

**Comment by sanity** [comment]
Ah, that's an issue, ok - I was under the impression that all <Input> elements kept their "value" in value, but obviously not so for checkboxes.

We need to extend the mechanism that reads the element value so that it can also read "checked", perhaps in the process generalizing it so it can read other values from elements also.

It shouldn't be too difficult, I'll try to get to it soon.

**Comment by sanity** [comment]
input().checked() will now return a `KVar<Boolean>` that will be synced to whether the input is checked. Will release `0.10.5` now with this improvement.

**Comment by geomagilles** [comment]
Hi - I had similar question with <select> element - not sure how to retrieve its current value

**Comment by sanity** [comment]
@geomagilles SelectElement implements [ValueElement](https://github.com/kwebio/kweb-core/blob/6939e53f90a1f98d381db81074e00600ef9df93c/src/main/kotlin/kweb/prelude.kt#L424) so you should be able to use getValue() and also the `value` property, do these not work?

We should add support for [selectedIndex](https://www.w3schools.com/jsref/prop_select_selectedindex.asp), this could mirror how [checked()](https://github.com/kwebio/kweb-core/blob/6939e53f90a1f98d381db81074e00600ef9df93c/src/main/kotlin/kweb/prelude.kt#L313) is implemented.

**Comment by geomagilles** [comment]
@sanity I did not manage to get it work - maybe only an example is missing

**Comment by sanity** [comment]
Hmm, yeah, something isn't right. This seems to work on an initial change but not on subsequent changes - it might relate to @Derek52's recent work on diffing value changes. Needs more investigation.

```kotlin
class SelectValueTestApp {

    internal lateinit var selectValue : KVar<String>

    val server: Kweb = Kweb(port= 7668) {
        doc.body {
            val select = select(name = "pets") {
                option().setAttribute("value", "dog").text("Dog")
                option().setAttribute("value", "dog").text("Cat")
            }
            selectValue = select.value
            selectValue.addListener { old, new ->
                println("$old -> $new")
            }
        }
    }

}
```

**Comment by geomagilles** [comment]
> it might relate to @Derek52's recent work on diffing value changes.

I had the issue using last release

**Comment by sanity** [comment]
Should be fixed in `0.1.6`.
